src/packagedcode/golang.py

```diff
@@ -0,0 +1,139 @@
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# When you publish or redistribute any data created with ScanCode or any ScanCode
+# derivative work, you must accompany this data with the following acknowledgment:
+#
+#  Generated with ScanCode and provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+#  OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+#  ScanCode should be considered or used as legal advice. Consult an Attorney
+#  for any legal advice.
+#  ScanCode is a free software code scanning tool from nexB Inc. and others.
+#  Visit https://github.com/nexB/scancode-toolkit/ for support and download.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from collections import OrderedDict
+import io
+import logging
+import re
+
+import attr
+from packageurl import PackageURL
+
+from commoncode import filetype
+from commoncode import fileutils
+from packagedcode.go_mod import GoMod
+from packagedcode import models
+
+
+"""
+Handle Go packages including go.mod and go.sum files.
+"""
+"""
+Sample go.mod file:
+module github.com/alecthomas/participle
+require (
+	github.com/alecthomas/repr v0.0.0-20181024024818-d37bc2a10ba1
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/stretchr/testify v1.4.0
+)
+go 1.13
+"""
+
+# TODO:
+# go.mod file does not contain version number.
+# valid download url need version number
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class GolangPackage(models.Package):
+    metafiles = ('go.mod',)
+    default_type = 'golang'
+    default_primary_language = 'Go'
+    default_web_baseurl = 'https://pkg.go.dev'
+    default_download_baseurl = None
+    default_api_baseurl = None
+
+    @classmethod
+    def recognize(cls, location):
+        if fileutils.file_name(location).lower() == 'go.mod':
+            gomod_obj = GoMod()
+            gomod_data = gomod_obj.parse_gomod(location)
+            yield build_gomod_package(gomod_data)
+
+    @classmethod
+    def get_package_root(cls, manifest_resource, codebase):
+        return manifest_resource.parent(codebase)
+
+    def repository_homepage_url(self):
+        return self.homepage_url
+
+
+def build_gomod_package(gomod_data):
+    """
+    Return a Package object from a go.mod file or None.
+    """
+
+    package_dependencies = []
+    require = gomod_data.get('require') or []
+    for name, version in require:
+        package_dependencies.append(
+            models.DependentPackage(
+                purl=PackageURL(
+                    type='golang',
+                    name=name
```

Comment:
is there any namespace in there?

Reply:
The module section contains the name of the package like `kingpin` is the name in `module github.com/alecthomas/kingpin`. You can check go.mod files which I have added and other files too, I think this will be the name.

Reply:
The module section contains the name of the package like kingpin is the name in module github.com/alecthomas/kingpin. You can check go.mod files which I have added and other files too, I think this will be the name.

Reply:
would  `github.com/alecthomas` be the namespace?

Reply:
I think whole `github.com/alecthomas/kingpin` will come in name as we can see in https://pkg.go.dev/

Reply:
I get that, I am talking about from a Package URL point of view. What is the name? and what is the namespace?

Reply:
@pombredanne I think in `github.com/alecthomas/kingpin`, `github.com/alecthomas` will be namespace and `kingpin` will be name

Reply:
@rpotter12 See @TG1999 's function here: https://github.com/package-url/packageurl-python/blob/5f6f7bb14279c11593003dc93ed1e46bf48cc9cb/src/packageurl/contrib/url2purl.py#L360

Reply:
This function converts a github url to a PackageURL Object, which has `name` and `namespace` values filled in correctly.